src/DynamoCore/Core/DynamoMigrator.cs

```diff
@@ -149,5 +270,143 @@ public static IEnumerable<FileVersion> GetInstalledVersions(string rootFolder)
 
             return fileVersions;
         }
+
+        /// <summary>
+        /// Given a root user data directory, this migrates user preferences,
+        /// packages and custom node definition files from one installed version
+        /// into another
+        /// </summary>
+        /// <param name="userDataDir"> root user data directory </param>
+        /// <param name="fromVersion"> source Dynamo version from which to migrate </param>
+        /// <param name="toVersion"> target Dynamo version into which to migrate </param>
+        /// <returns> preference settings read from preference settings file </returns>
+        internal static PreferenceSettings Migrate(string userDataDir, FileVersion fromVersion, FileVersion toVersion)
+        {
+            // Create concrete DynamoMigratorBase object using previousVersion and reflection
+            var sourceMigrator = CreateMigrator(userDataDir, fromVersion);
+            Debug.Assert(sourceMigrator != null);
+
+            // get migrator object for current version
+            var targetMigrator = CreateMigrator(userDataDir, toVersion);
+            Debug.Assert(targetMigrator != null);
+            var preferenceSettings = targetMigrator.WritePreferences(sourceMigrator);
+
+            targetMigrator.MigratePackages(sourceMigrator);
+
+            targetMigrator.MigrateDefinitions(sourceMigrator);
+
+            return preferenceSettings;
+        }
```

Comment:
This is a huge step forward, good stuff! It's just that it's not nearly far enough. As discussed we'd ideally like to have file copy related stuff all captured inside each migrator class. Imagine what we will have to do here, if we choose to copy over `Logs` directory:

1. You'll have to introduce a new method `DynamoMigratorBase.MigrateLogs`.
2. You'll have to update this `Migrate` API.
3. You'll have to add virtual methods to N-number of `DynamoMigratorBase` derived classes.

Instead of doing that, I'm expecting the `Migrate` method to look this way:

```csharp
internal static PreferenceSettings Migrate(string userDataDir, FileVersion fromVersion, FileVersion toVersion)
{
    var sourceMigrator = CreateMigrator(userDataDir, fromVersion);
    var targetMigrator = CreateMigrator(userDataDir, toVersion);
    
    targetMigrator.MigrateFrom(sourceMigrator);
    return targetMigrator.PreferenceSettings;
}
```

In fact, it also feels unusual for this method to return `PreferenceSettings` -- why does it receive such special treatment? It should have been like any other data we include in migration. Perhaps the call can return the target `DynamoMigratorBase`, and then leave it up to the caller to determine what they need to do with it. This `DynamoMigratorBase` will travel back all the way into `DynamoModel` which owns `PreferenceSettings`, then `DynamoModel` will update its internal copy of `PreferenceSettings` with the one from this `targetMigrator`.

Does that make sense?